File: leetcode/problems/cpp/binary-tree-preorder-traversal.cpp
// by @codeAbinash
// Time : O(n)
// Space : O(n)

#include "vector"
using namespace std;

struct TreeNode {
   int val;
   TreeNode* left;
   TreeNode* right;
   TreeNode() : val(0), left(nullptr), right(nullptr) {}
   TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
   TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};

class Solution {
   vector<int> result;
   void preorder(TreeNode* node) {
      if (!node) return;
      result.push_back(node->val);
      preorder(node->left);
      preorder(node->right);
   }
public:
   vector<int> preorderTraversal(TreeNode* root) {
      preorder(root);
      return result;
   }
};

File: leetcode/problems/ts/binary-tree-preorder-traversal.ts
// by @codeAbinash
// Time : O(n)
// Space : O(n)

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function dfs(root: TreeNode | null, ans: number[]) {
  if (!root) return;
  ans.push(root.val);
  dfs(root.left, ans);
  dfs(root.right, ans);
}
function preorderTraversal(root: TreeNode | null): number[] {
  const ans: number[] = [];
  dfs(root, ans);
  return ans;
}
